### ISSUE TITLE
OpenCL: Reduce binary size, bitmaps and hash tables to 64-bit for more formats

### ISSUE THREAD

**Issue by magnumripper** [openwall/john#5257]
First out is raw MD4 and MD5, which share code with NT (done in #5246). These two are also identical enough to each other that the patch for fixing one possibly applies cleanly to the other.

Then there are eight more with code copies in two or more slightly different versions. Perhaps they too can use shared code instead.

**Comment by solardiz** [comment]
Will we still be doing full comparisons in `cmp_exact`? Will we keep the full source (ASCII) for that? A way to save more memory is the opposite, to drop ASCII source and reconstruct it from binary for writing into pot file. So maybe we want to reduce the binary size as transferred to OpenCL device only, not binary size as maintained by host?

On host, maybe we want to separately store hot vs. cold parts of binaries, which core currently does not support - the approach so far was that the (partial) binaries are hot, and the (full) sources are cold, but then we decided that it wastes too much memory to always keep the ASCII sources.

**Comment by solardiz** [comment]
> On host, maybe we want to separately store hot vs. cold parts of binaries

OTOH, the extra indirection could cost just as much memory - e.g., a half of an NTLM hash is 64 bits, and a pointer on a modern system is also 64 bits. So this is only worthwhile for larger binaries or if we avoid the need to store separate pointers to the cold parts - e.g., we allocate them such that their addresses can be calculated from those of the hot parts, which can be done by interleaving e.g. 2 MiB blocks of hots and colds (2 MiB being large enough that the colds don't commonly fall into the same cache lines and TLB entries with the hots, nor waste ranges of cache tags needed for the hots).

**Comment by solardiz** [comment]
A different angle is that when using OpenCL with hash "comparisons" implemented on device, the hot vs. cold split on host is unimportant. Saving host memory can still be important, so where we aren't too lazy we can keep full binaries and reconstruct source for writing into pot. We can also use partial binaries for/in the data structures transferred to device - so I assume that's the task here.

Just brainstorming here, without having yet reviewed the changes @magnumripper already implemented for NT - sorry!

**Comment by magnumripper** [comment]
> Will we still be doing full comparisons in `cmp_exact`? Will we keep the full source (ASCII) for that?

Yes, and currently we will warn loudly about false hits (or 64-bit collisions if that would happen).

> half of an NTLM hash is 64 bits, and a pointer on a modern system is also 64 bits

It might be that the source always contains `$NT$` after `prepare()` or `split()` (I didn't look now), making it more than 64 bits. OTOH that could be fixed by changing prepare/split (it could even be regarded a bug! It should be the other way round). Anyway, it would be a separate issue and PR before this gets out of control :)

**Comment by magnumripper** [comment]
After doing this with NT, I have come to realize there isn't that much performance to gain, as the performance is mostly dictated by the bitmap - which basically have the same characteristics when using a 64-bit binary.

For raw formats in particular, the memory saving itself is still interesting of course. The best overall savings would be full binaries on host side and a `source()` function, and partial binaries on GPU side. Even NT doesn't have the former yet - my work so far was basically GPU side.
